[SelectOpt] Add handling for Select-like operations.

davemgreen opened the pull request:
Some operations behave like selects. For example `or(zext(c), y)` is the same as `select(c, y|1, y)` and instcombine can canonicalize the select to the or form. These operations can still be worthwhile converting to branch as opposed to keeping as a select or or instruction.

This patch attempts to add some basic handling for them, creating a SelectLike abstraction in the select optimization pass. The backend can opt into handling `or(zext(c),x)` as a select if it could be profitable, and the select optimization pass attempts to handle them in much the same way as a `select(c, x|1, x)`. The Or(x, 1) may need to be added as a new instruction, generated as the or is converted to branches.

This helps fix a regression from selects being converted to or's recently.

davemgreen commented:
I've pushed a rebase to try and move away from the failing CI tests.

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.h:
```diff
@@ -412,6 +412,16 @@ class AArch64TTIImpl : public BasicTTIImplBase<AArch64TTIImpl> {
+    // break point in the code - the end of a block with an unconditional
+    // terminator.
+    return isa<SelectInst>(I) ||
+           (isa<BinaryOperator>(I) && isa<BranchInst>(I->getNextNode()) &&
```
Is the next node guaranteed to exist in all cases? I assume a well-formed block should always have a terminator, but I wasn't completely sure. If the next node could be NULL then you'd have to add an extra check before calling `isa`.

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.h:
```diff
@@ -412,6 +412,16 @@ class AArch64TTIImpl : public BasicTTIImplBase<AArch64TTIImpl> {
+    // terminator.
+    return isa<SelectInst>(I) ||
+           (isa<BinaryOperator>(I) && isa<BranchInst>(I->getNextNode()) &&
+            cast<BranchInst>(I->getNextNode())->isUnconditional());
```
Also, just for my understanding why do we only care about treating instructions like select if they are at the end of the block? From the name of the function `shouldTreatInstructionLikeSelect` it sounds like we should always treat `or(zext(c),y` like a select. I guess the function is really asking the question - "should we convert this like select-like instruction into compare-and-branch?". So I sympathise with you and it's hard to put that mouthful into a succint function name!

david-arm reviewed (commented):
Just a couple of comments so far, but will review the SelectOptimize code soon as well!

david-arm commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -129,6 +124,151 @@ class SelectOptimizeImpl {
+  private:
+    SelectLike(Instruction *SI) : SI(SI) {}
+
+    Instruction *SI;
```
The name, and the routines below (getSI, etc.) do suggest that this is a SelectInst so could that be a little confusing? I don't know what the alternative would be though - perhaps SLI (Select-Like Instruction) or even just `I`?

david-arm commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -368,15 +508,24 @@ void SelectOptimizeImpl::optimizeSelectsInnerLoops(Function &F,
+  if (isa<BinaryOperator>(SI.getSI())) {
+    V = SI.getFalseValue();
+    if (isTrue)
+      V = IB.CreateOr(V, ConstantInt::get(V->getType(), 1));
```
Is this assuming there is currently on one type of BinaryOperator that is Select-like, i.e. an `or` instruction? If so, should that logic be hidden away in the SelectLike class?

david-arm commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -129,6 +124,151 @@ class SelectOptimizeImpl {
+      // is a logical-and/logical-or then it is better treated as a and/or by
+      // the backend.
+      if (isa<SelectInst>(I) &&
+          !PatternMatch::match(I,
```
I'm not sure if I've missed something here, but if `I` is a SelectInst how could it pattern match to a `or` or an `and` instruction? It seems like we'll never match.

Also, this type of logic feels like it could live in `shouldTreatInstructionAsSelect`?

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.h:
```diff
@@ -412,6 +412,16 @@ class AArch64TTIImpl : public BasicTTIImplBase<AArch64TTIImpl> {
+    // selects, here we only transform them if they are already at a natural
+    // break point in the code - the end of a block with an unconditional
+    // terminator.
+    return isa<SelectInst>(I) ||
```
Instead of checking for `BinaryOperator` here, do we actually want to explicitly look for `or` instructions instead? I'm just wondering what happens if later on someone extends the SelectLike class in SelectOptimize to cover more binary operators.

I actually wonder if this function should be called directly by SelectLike::match and whether that simplifies the logic a little?

davemgreen committed:
[SelectOpt] Add handling for Select-like operations.

Some operations behave like selects. For example `or(zext(c), y)` is the same
as select(c, y|1, y)` and instcombine can canonicalize the select to the or
form. These operations can still be worthwhile converting to branch as opposed
to keeping as a select or Or instruction.

This patch attempts to add some basic handling for them, creating a SelectLike
abstraction in the select optimization pass. The backend can opt into handling
`or(zext(c),x)` as a select if it could be profitable, and the select
optimization pass attempts to handle them in much the same way as a
`select(c, x|1, x)`. The Or(x, 1) may need to be added as a new instruction,
generated as the or is converted to branches.

This helps fix a regression from selects being converted to or's recently.

sapostolakis reviewed (commented):
Thanks. This sounds like a good idea.
Left some comments. 
Main issue is that I would like to have a flag that could disable, if needed, the select-like (but not proper select) part. The rest of the comments are nits.

sjoerdmeijer commented:
Thanks for working on this! I will take this patch, see how it behaves on the Neoverse V2, and will let you know soon.

sjoerdmeijer commented:
This improves MCF from SPEC INT 2017 by more than 9%, which is great!

davemgreen commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -378,6 +378,15 @@ class TargetTransformInfoImplBase {
 
   bool enableSelectOptimize() const { return true; }
 
+  bool shouldTreatInstructionLikeSelect(Instruction *I) {
```
The new select-like `or(zext(c), x)` is currently only enabled for AArch64. I haven't verified the performance on any other systems. The default is still currently just selects, minus the "logical-and/or" from https://reviews.llvm.org/D138490. I think that adding the option to disable the new or(zext) sounds like a good idea, I've added it for AArch64. Let me know if you thing this should add another option to enable it more generally in this function too.

sapostolakis replied:
thanks. adding EnableOrLikeSelectOpt for AArch64 is good enough for me.

davemgreen committed:
Const correctness changes and AArch64 option

sapostolakis commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -1023,8 +1176,8 @@ SmallPtrSet<const Instruction *, 2>
   for (const SelectGroup &ASI : SIGroups)
-    for (const SelectInst *SI : ASI)
-      SIset.insert(SI);
+    for (SelectLike SI : ASI)
```
fair point. given that this is just a pointer it is fine. In my mind, I thought of it as a Instruction and this is why it seemed necessary to use references.

sapostolakis commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -945,7 +1103,7 @@ bool SelectOptimizeImpl::computeLoopCosts(
     CostInfo &MaxCost = LoopCost[Iter];
     for (BasicBlock *BB : L->getBlocks()) {
-      for (const Instruction &I : *BB) {
+      for (Instruction &I : *BB) {
```
I see. I am okay with either way. Just wanted to make sure it was not removed by mistake.

sapostolakis commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -1038,7 +1191,7 @@ SelectOptimizeImpl::computeInstCost(const Instruction *I) {
 
 ScaledNumber<uint64_t>
-SelectOptimizeImpl::getMispredictionCost(const SelectInst *SI,
+SelectOptimizeImpl::getMispredictionCost(SelectLike SI,
```
ignore the reference point as discussed in the other comment but for the const part, could we keep it here?

sapostolakis commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -378,6 +378,15 @@ class TargetTransformInfoImplBase {
+    // If the select is a logical-and/logical-or then it is better treated as a
+    // and/or by the backend.
+    using namespace llvm::PatternMatch;
+    return isa<SelectInst>(I) &&
```
the added EnableOrLikeSelectOpt resolved my issue here

davemgreen commented on llvm/lib/CodeGen/SelectOptimize.cpp:
```diff
@@ -129,6 +124,146 @@ class SelectOptimizeImpl {
+      // An Or(zext(i1 X), Y) can also be treated like a select, with condition
+      // C and values Y|1 and Y.
+      Value *X;
+      if (PatternMatch::match(
```
Yep, It needs to not match the match() function in this class.

davemgreen reviewed (commented):
> I might be wrong, but I suspect we could do with some more test-cases, especially negative tests when this shouldn't trigger. From a quick look, things like when the pattern doesn't exactly matches, one use, dbg instructions, if it is not a terminator, etc.

There should be tests for one uses select groups and the positions in selectopt.ll. See 365fbbfbcfefb8766f7716109b9c3767b58e6058. Debug info tests we would not usually add unless we had good reason, they can be quite bulky, but I've added a run line. Anything else specific you think needs adding?

davemgreen committed:
More const's and a debug run line

sjoerdmeijer commented:
> > I might be wrong, but I suspect we could do with some more test-cases, especially negative tests when this shouldn't trigger. From a quick look, things like when the pattern doesn't exactly matches, one use, dbg instructions, if it is not a terminator, etc.
> 
> There should be tests for one uses select groups and the positions in selectopt.ll. See [365fbbf](https://github.com/llvm/llvm-project/commit/365fbbfbcfefb8766f7716109b9c3767b58e6058). Debug info tests we would not usually add unless we had good reason, they can be quite bulky, but I've added a run line. Anything else specific you think needs adding?

Ok, thanks for clarifying, looks fine then.

sjoerdmeijer reviewed (approved):
Looks like a good change to me, but give it a few days in case there are more comments.

sapostolakis commented:
LGTM as well. No further comments from me.

davemgreen committed:
Fixup after rebase
